### PR TITLE
[WIP]  switchDetection2D, merge detection2D programs #79

### PR DESCRIPTION
**Pull request by davidvelascogarcia**
# [vision] switchDetection2D

Addresses https://github.com/roboticslab-uc3m/vision/issues/79.

## Description
Pull request adds `switchDetection2D` module. `switchDetection2D` module, merge [vision](https://github.com/roboticslab-uc3m/vision) detection2D programs. The module merge [colorRegionDetection2D](https://github.com/roboticslab-uc3m/vision/tree/develop/programs/colorRegionDetection2D), [haarDetection2D](https://github.com/roboticslab-uc3m/vision/tree/develop/programs/haarDetection2d) and [tensorflowDetection2D](https://github.com/roboticslab-uc3m/vision/tree/develop/programs/tensorflowDetection2D). This module works with `segmentorThread`.  Using [vision](https://github.com/roboticslab-uc3m/vision) detection2D programs structure. 

Choosing detection mode running the program with `--switchMode` context, selecting:
* colorDetection
* haarDetection
* tensorflowDetection

It allows previous configuration parameters of diferent [vision](https://github.com/roboticslab-uc3m/vision) detection2D programs, like `--algorithm` in colorRegion, `--xmlFile` in haar or `--pbTrainedModel` in tensorflow.

Default [vision](https://github.com/roboticslab-uc3m/vision) detection2D program is `haarDetection2D`.
The process diagram attached in Figure 1.

swicthDetection
Figure 1. Diagram

Each previous [vision](https://github.com/roboticslab-uc3m/vision) detection2D programs, it´s created like a class. `switchDetection2D` and `segmentorThread::run` build the object of the selected detection mode. The detector mode selected receive `yarp image`, process the image and detect the features, and return the processed image and results. After that processed image and results are published with `yarp` ports.

### CODE REVIEW

**Comment by davidvelascogarcia** [comment]
Speaking with @jgvictores, i´m going to try to add [YarpPlugins/RealToSimControlboard/Transformation](https://github.com/roboticslab-uc3m/tools/blob/develop/libraries/YarpPlugins/RealToSimControlboard/Transformation.hpp) class to build the detector objects with parameters objects instead of each parameter individually.

File: programs/switchDetection2D/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+    message(WARNING "OpenCV package not found, disabling switchDetection2D program")
+endif()
+
+if(NOT TensorflowCC_FOUND AND (NOT DEFINED ENABLE_switchDetection2D OR ENABLE_switchDetection2D))
+    message(WARNING "TensorflowCC package not found, disabling switchDetection2D program")
+endif()
+
+
+cmake_dependent_option(ENABLE_switchDetection2D "Enable/disable switchDetection2D program" ON
+"ENABLE_TravisLib;OpenCV_FOUND;TensorflowCC_FOUND" OFF)
```
Comment (PeterBowman) [inline programs/switchDetection2D/CMakeLists.txt]: It is my understanding that this app requires all three components listed on these lines. I'm especially concerned about the Tensorflow dependency. I don't need it in order to use the *haarDetection2D* program, but it is a blocker for me if *switchDetection2D* is meant to serve as a full replacement. Thus, all Tensorflow-related source files in this targets need to be enabled conditionally on the CMake level (via `if` clauses), and macro guards be added to the cpp in which selection is performed on runtime between the three embedded detectors.

On a side note, I guess the *colorRegionDetection2D* and *tensorflowDetection2D* apps will disappear (they're still there on current source directory, whereas *haarDetection2D* has been already replaced).

**Comment by jgvictores** [comment]
> Speaking with @jgvictores, i´m going to try to add [YarpPlugins/RealToSimControlboard/Transformation](https://github.com/roboticslab-uc3m/tools/blob/develop/libraries/YarpPlugins/RealToSimControlboard/Transformation.hpp) class to build the detector objects with parameters objects instead of each parameter individually.

Well, a similar class inspired on `Transformation`. : )

File: programs/switchDetection2D/CMakeLists.txt (hunk 1 above)
Comment (jgvictores) [reply programs/switchDetection2D/CMakeLists.txt]: Marked PR as WIP. I expect (or at least hope :blush:) work on what I commented at https://github.com/roboticslab-uc3m/vision/pull/98#issuecomment-558532769 will enable enough polymorphism to be able to load components that are compiled as separate target libraries with clearly differentiated dependencies.

Comment (PeterBowman) [reply programs/switchDetection2D/CMakeLists.txt]: Embedding is also an option, though :) (https://github.com/roboticslab-uc3m/vision/issues/79#issuecomment-552521792).

**Comment by jgvictores** [comment]
This got way too messy. Going to re-do towards https://github.com/roboticslab-uc3m/vision/issues/79 in a new branch. Ciao!

**Comment by PeterBowman** [comment]
Reimplemented by @jgvictores and merged at https://github.com/roboticslab-uc3m/vision/commit/7497b4b07688aafa101889f5285abddc1c16c427.

**Pull request closed without merging**